unicorn_mode/build_unicorn_support.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
+#
+# By Ziqiao Kong <mio@lazym.io>
+
+import subprocess
+import sys
+import shutil
+import os
+import tempfile
+import json
+from pathlib import Path
+
+# https://stackoverflow.com/questions/1871549/how-to-determine-if-python-is-running-inside-a-virtualenv
+def in_venv():
+    return sys.prefix != sys.base_prefix
+
+def run_cmd(cmd: str, cwd: Path = None, quiet: bool = False):
+    if not cwd:
+        cwd = Path(__file__).parent
+    if quiet:
+        print(f"[*] Runing queitly: \"{cmd}\" under workding directory {cwd}")
+    else:
+        print(f"[*] Runing: \"{cmd}\" under workding directory {cwd}")
+    if quiet:
+        try:
+            out = subprocess.check_output(cmd, shell=True, stderr=subprocess.PIPE, cwd=cwd)
+        except subprocess.CalledProcessError as e:
+            print(f"[!] Failed with:\n{e.stdout}\n{e.stderr}")
+            raise e
+        return out
+    else:
+        subprocess.check_call(cmd, shell=True, cwd=cwd)
+        return None
+
+
+def detect_from_env_or_file(target: str):
+    if target in os.environ:
+        return os.environ[target]
+    elif (cwd / target).exists():
+        with open(cwd / target, mode="r") as f:
+            return f.read().strip()
+    else:
+        return None
+
+cwd = Path(__file__).parent
+
+libs_path = cwd / "lib"
+include_path = cwd / "include"
+unicornafl_path = cwd / "unicornafl"
+
+if libs_path.exists():
+    print("[!] Cleaning previous artifacts...")
+    shutil.rmtree(libs_path)
+if include_path.exists():
+    print("[!] Cleaning previous headers...")
+    shutil.rmtree(include_path)
+
+if not (cwd.parent / "afl-showmap").exists():
+    print("[!] Please compile AFL++ first.")
+    exit(1)
+
+if not shutil.which("cargo"):
+    print("[!] No cargo, please install Rust in advance.")
+    print("[!] TLDR: `curl https://sh.rustup.rs -sSf | sh -s -- -y`")
+    exit(1)
+
+unicornafl_version = detect_from_env_or_file("UNICORNAFL_VERSION")
+if not unicornafl_version:
+    print("[!] No valid UNICORNAFL_VERSION found")
+    exit(1)
+
+try:
+    run_cmd("git status", cwd)
+except subprocess.CalledProcessError:
+    print(f"[!] Unicornafl is not a submodule, do a separate clone...")
+    run_cmd("rm -rf unicornafl && git clone https://github.com/AFLplusplus/unicornafl", cwd)
+
+if not (unicornafl_path / ".git").exists():
+    print(f"[!] Submodule not existing, will do a checkout first")
+    run_cmd("git submodule update --init --recursive", cwd)
+
+print(f"[*] We will checkout unicornafl {unicornafl_version}")
+run_cmd(f"git fetch --all && git checkout {unicornafl_version}", unicornafl_path)
+
+print(f"[*] Now building unicornafl python bindings")
+venv = in_venv()
+if not venv:
+    print(f"[!] A python venv is highly recommended! We will try to add --user for you.")
+
+if not shutil.which("maturin"):
+    print(f"[!] No maturin, will install maturin first")
+    if not venv:
+        run_cmd("pip install --user maturin")
```

Comment:
shouldn't this possibly be `python3 -m pip`? At least on some distributions(?)

Reply:
This is actually tricky. For example, on `manylinux`, `python3 -m pip` will definitely be rejected and the expected usages are like `python3.10 -m pip` or `python3.11 -m pip` or `pipx`. Generally, both `pip` and `python3 -m pip` are not correct and I would say we should not install anything if a venv is not present nowadays, if that is acceptable.